src/main.c

```diff
@@ -0,0 +1,257 @@
+       (events[0].events & EPOLLIN) > 0) {
+      fprintf(stderr,
+              "ClientSession wants us to read_tls. First we need to pull some "
+              "bytes from the socket\n");
+
+      memset(buf, 0, sizeof(buf));
+      n = read(sockfd, buf, sizeof(buf));
+      if(n == 0) {
+        fprintf(stderr, "EOF reading from socket\n");
+        break;
+      }
+      else if(n < 0) {
+        perror("reading from socket");
+        return 1;
+      }
+      fprintf(stderr, "read %d bytes from socket\n", n);
+
+      /*
+       * Now pull those bytes from the buffer into ClientSession.
+       * Note that we pass buf, n; not buf, sizeof(buf). We don't
+       * want to pull in unitialized memory that we didn't just
+       * read from the socket.
+       */
+      n = rustls_client_session_read_tls(client_session, buf, n);
+      if(n == 0) {
+        fprintf(stderr, "EOF from ClientSession::read_tls\n");
+        // TODO: What to do here?
+        break;
+      }
+      else if(n < 0) {
+        fprintf(stderr, "Error in ClientSession::read_tls\n");
+        return 1;
+      }
+
+      int result = rustls_client_session_process_new_packets(client_session);
+      if(result != CRUSTLS_OK) {
+        fprintf(stderr, "Error in process_new_packets");
+        return 1;
+      }
+
+      /* Read all available bytes from the client_session until EOF.
+       * Note that EOF here indicates "no more bytes until
+       * process_new_packets", not "stream is closed".
+       */
+      for(;;) {
+        memset(buf, 0, sizeof(buf));
+        n = rustls_client_session_read(client_session, buf, sizeof(buf));
+        if(n == 0) {
+          fprintf(stderr, "EOF from ClientSession::read (this is expected)\n");
+          break;
+        }
+        else if(n < 0) {
+          fprintf(stderr, "Error in ClientSession::read\n");
+          return 1;
+        }
+
+        write_all(STDOUT_FILENO, buf, n);
+      }
+    }
+    if(rustls_client_session_wants_write(client_session) &&
+       (events[0].events & EPOLLOUT) > 0) {
+      fprintf(stderr, "ClientSession wants us to write_tls.\n");
+      memset(buf, 0, sizeof(buf));
+      n = rustls_client_session_write_tls(client_session, buf, sizeof(buf));
+      if(n == 0) {
+        fprintf(stderr, "EOF from ClientSession::write_tls\n");
+        return 1;
+      }
+      else if(n < 0) {
+        fprintf(stderr, "Error in ClientSession::write_tls\n");
+        return 1;
+      }
+
+      write_all(sockfd, buf, n);
+    }
+  }
+
+  return 0;
+}
+
+int
+main(int argc, const char **argv)
+{
+  if(argc <= 2) {
+    fprintf(stderr,
+            "usage: %s hostname path\n\n"
+            "Connect to a host via HTTPS on port 443, make a request for the\n"
+            "given path, and emit response to stdout.\n",
+            argv[0]);
+    return 1;
+  }
+  const char *hostname = argv[1];
+  const char *path = argv[2];
+
+  const void *client_config = rustls_client_config_new();
```

Comment:
This "leaks" in any of the error cases below because you `return 1` without a call to `rustls_client_config_free`. I put "leaks" in quotes because obviously the process existing is pretty effective garbage collection. I'd suggest putting a `cleanup` label at the bottom of functions that allocate resources. Then you can do:

```
  int ret = -1;
  void *must_be_freed = some_function();

  if (!do_something_else(must_be_freed)) {
    goto cleanup;
  }

  <more logic>

  // if we made it here, we succeeded
  ret = 0;

cleanup:
  free(must_be_freed);
  <free/close/release other resources>
  return ret;
```